src/core/thread/mle_router.cpp

```diff
@@ -4482,6 +4483,79 @@ Error MleRouter::SendTimeSync(void)
 }
 #endif // OPENTHREAD_CONFIG_TIME_SYNC_ENABLE
 
+Error MleRouter::DetachGracefully(otDetachGracefullyCallback aCallback, void *aContext)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(aCallback != nullptr, error = kErrorInvalidArgs);
+
+    if (IsChild())
+    {
+        ExitNow(error = Mle::DetachGracefully(aCallback, aContext));
+    }
+
+    OT_ASSERT(IsDisabled() || IsDetached() || IsRouter() || IsLeader());
+
+    if (mDetachGracefullyCallback != nullptr)
+    {
+        ExitNow(error = kErrorBusy);
+    }
+
+    mDetachGracefullyCallback = aCallback;
+    mDetachGracefullyContext  = aContext;
+
+    if (IsLeader())
+    {
+        Stop();
+        mDetachGracefullyTimer.Start(0);
+        ExitNow(error = kErrorNone);
+    }
+    else if (IsDetached() || IsDisabled())
+    {
+        mDetachGracefullyTimer.Start(0);
+        ExitNow(error = kErrorNone);
+    }
+    else
+    {
+        SendAddressRelease(&MleRouter::HandleDetachGracefullyAddressReleaseResponse, this);
+        mDetachGracefullyTimer.Start(kDetachGracefullyTimeout);
+        ExitNow(error = kErrorNone);
+    }
+
+exit:
+    return error;
+}
+
+void MleRouter::HandleDetachGracefullyAddressReleaseResponse(void *               aContext,
+                                                             otMessage *          aMessage,
+                                                             const otMessageInfo *aMessageInfo,
+                                                             Error                aResult)
+{
+    static_cast<MleRouter *>(aContext)->HandleDetachGracefullyAddressReleaseResponse(aMessage, aMessageInfo, aResult);
+}
+
+void MleRouter::HandleDetachGracefullyAddressReleaseResponse(otMessage *          aMessage,
+                                                             const otMessageInfo *aMessageInfo,
+                                                             Error                aResult)
+{
+    OT_UNUSED_VARIABLE(aMessage);
+    OT_UNUSED_VARIABLE(aMessageInfo);
+    OT_UNUSED_VARIABLE(aResult);
+
+    if (mDetachGracefullyCallback != nullptr)
+    {
+        otDetachGracefullyCallback callback = mDetachGracefullyCallback;
+        void *                     context  = mDetachGracefullyContext;
+
+        Stop();
+        mDetachGracefullyTimer.Stop();
```

Comment:
```suggestion
        Stop();
```
should we move `mDetachGracefullyTimer.Stop();` to Stop() ?

Reply:
Moreover, detaching gracefully usually takes time. It's still possible that the user may stop Thread via other means during the graceful delay. In such case, should we also stop the gradeful timer and call the callback when Thread is stopped?

I think we need to make sure that the graceful attaching process is finalized if the Thread state transfer to `disabled` by any possible means.

Reply:
Agreed. Some things to mention:

1. We cannot synchronously call the callback in `Stop()`.
2. `Stop()` is the only place where the state can transition to `disabled`.

So instead of stopping the timer in `Stop()`, I made it start the timer with 0 delay, and let the timer stop itself. PTAL.

Reply:
> We cannot synchronously call the callback in Stop().

Any specific reason why?

Reply:
> > We cannot synchronously call the callback in Stop().
> 
> Any specific reason why?

I think calling callbacks synchronously should be generally avoided. An example here is that the CLI module callback is called before the command handler returns, causing `OutputResult` being called twice during the command handler.